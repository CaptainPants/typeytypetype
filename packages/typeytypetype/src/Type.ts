import { Definition } from './definitions/Definition.js';
import { ArrayDefinition } from './definitions/ArrayDefinition.js';
import { NumberTypeDefinition } from './definitions/NumberTypeDefinition.js';
import { StringTypeDefinition } from './definitions/StringTypeDefinition.js';
import { ConstantDefinition } from './definitions/ConstantDefinition.js';
import { BooleanTypeDefinition } from './definitions/BooleanTypeDefinition.js';
import { UnionDefinition } from './definitions/UnionModel.js';
import { IntersectionDefiniton } from './definitions/IntersectionDefiniton.js';
import { ObjectDefinition } from './definitions/ObjectDefinition.js';
import { MappedDefinition } from './internal/utilityTypes.js';
import { NamedDefinition } from './definitions/NamedDefinition.js';

export const Type = {
    constant<TValue extends string | number | boolean>(
        value: TValue
    ): ConstantDefinition<TValue> {
        return new ConstantDefinition(value);
    },

    null() {
        return new ConstantDefinition(null);
    },

    undefined(): Definition<undefined> {
        return new ConstantDefinition(undefined);
    },

    string(): StringTypeDefinition {
        return new StringTypeDefinition();
    },

    boolean(): BooleanTypeDefinition {
        return new BooleanTypeDefinition();
    },

    number(): NumberTypeDefinition {
        return new NumberTypeDefinition();
    },

    union<TTypes extends readonly unknown[]>(
        ...definitions: MappedDefinition<TTypes>
    ) {
        return new UnionDefinition(definitions);
    },

    intersect<TTypes extends readonly unknown[]>(
        ...definitions: MappedDefinition<TTypes>
    ) {
        return new IntersectionDefiniton(definitions);
    },

    object<TPropertyTypes extends Record<string, unknown>>(
        propertyDefinitions: MappedDefinition<TPropertyTypes>
    ) {
        return new ObjectDefinition(propertyDefinitions);
    },

    array<TITemType>(
        itemDefinition: Definition<TITemType>
    ): ArrayDefinition<TITemType> {
        return new ArrayDefinition(itemDefinition);
    },

    named<TType>(name: string): NamedDefinition<TType> {
        return new NamedDefinition<TType>(name);
    },
};
