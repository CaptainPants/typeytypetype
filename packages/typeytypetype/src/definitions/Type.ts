import { type Definition } from './Definition.js';
import { ArrayDefinition } from './ArrayDefinition.js';
import { NumberDefinition } from './NumberDefinition.js';
import { StringDefinition } from './StringDefinition.js';
import {
    BooleanConstantDefinition,
    NullConstantDefinition,
    NumberConstantDefinition,
    StringConstantDefinition,
    UndefinedConstantDefinition,
} from './ConstantDefinition.js';
import { BooleanDefinition } from './BooleanDefinition.js';
import { UnionDefinition } from './UnionDefinition.js';
import { DeferredDefinition } from './DeferredDefinition.js';
import { RigidObjectDefinition } from './RigidObjectDefinition.js';
import { MapObjectDefinition } from './MapObjectDefinition.js';
import { type MappedDefinition } from './internal/types.js';

function constant(value: string): StringConstantDefinition;
function constant(value: number): NumberConstantDefinition;
function constant(value: boolean): BooleanConstantDefinition;
function constant(value: unknown): unknown {
    if (typeof value === 'string') {
        return new StringConstantDefinition(
            value
        ) satisfies Definition<string> as any;
    } else if (typeof value === 'number') {
        return new NumberConstantDefinition(
            value
        ) satisfies Definition<number> as any;
    } else if (typeof value === 'boolean') {
        return new BooleanConstantDefinition(
            value
        ) satisfies Definition<boolean> as any;
    } else {
        throw new TypeError('Not supported');
    }
}

export const Type = {
    constant,

    null() {
        return new NullConstantDefinition();
    },

    undefined(): Definition<undefined> {
        return new UndefinedConstantDefinition();
    },

    string(): StringDefinition {
        return new StringDefinition();
    },

    boolean(): BooleanDefinition {
        return new BooleanDefinition();
    },

    number(): NumberDefinition {
        return new NumberDefinition();
    },

    union<TTypes extends readonly unknown[]>(
        ...definitions: MappedDefinition<TTypes>
    ) {
        // making the type system unhappy here
        return new UnionDefinition<TTypes[number]>(definitions as any);
    },

    object<TObject extends Readonly<Record<string, unknown>>>(
        propertyDefinitions: MappedDefinition<TObject>
    ) {
        return new RigidObjectDefinition<TObject>(propertyDefinitions);
    },

    map<TValue>(valueDefinition: Definition<TValue>) {
        return new MapObjectDefinition<TValue>(valueDefinition);
    },

    array<TElement>(
        elementDefinition: Definition<TElement>
    ): ArrayDefinition<TElement> {
        return new ArrayDefinition(elementDefinition);
    },

    named<TType>(name: string): DeferredDefinition<TType> {
        return new DeferredDefinition<TType>(name);
    },
};
