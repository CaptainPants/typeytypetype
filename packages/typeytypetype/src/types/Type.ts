import {
    type ValidationOptions,
    type ValidationResult,
} from '../validation/types.js';

export interface Type<T> {
    matches: (value: unknown) => value is T;

    doMatches: (value: unknown, deep: boolean, depth: number) => value is T;

    validate: (value: unknown, options?: ValidationOptions) => ValidationResult;

    validateAndThrow: (
        value: unknown,
        options?: ValidationOptions
    ) => Promise<T>;

    doValidate: (
        value: unknown,
        options: ValidationOptions,
        depth: number
    ) => ValidationResult;

    toTypeString: () => string;

    doToTypeString: (depth: number) => string;

    hasLabel: (label: string) => boolean;

    getAttribute: (name: string) => unknown;
}
