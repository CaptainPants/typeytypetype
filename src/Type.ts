export interface Model<T> {
  metadata: unknown;
}

export type UnModel<TModel> = TModel extends Model<infer T> ? T : TModel;

type UnModelArray<T> = { [TKey in keyof T]: UnModel<T[TKey]> };

type MapModel<T> = {
    [TKey in keyof T]: Model<T[TKey]>;
}

type Intersect<TArgs extends readonly unknown[], TDepth extends readonly unknown[] = []> =
    TDepth['length'] extends 20 ? "Error" 
    : TArgs['length'] extends 1 ? TArgs[0]
    : TArgs extends [infer First, ...infer Rest] ? First & Intersect<Rest, [1, ...TDepth]> 
    : never;

export namespace Type {
  export function string<TString extends string = string>(
    values?: TString
  ): Model<TString> {
    throw "Not implemented";
  }

  export function union<TArgs extends readonly Model<unknown>[]>(
    ...models: TArgs
  ): Model<UnModelArray<TArgs>[number]> {
    throw "Not implemented";
  }

  export function intersect<TArgs extends readonly Model<unknown>[]>(
    ...models: TArgs
  ): MapModel<Intersect<UnModelArray<TArgs>>> {
    throw "Not implemented";
  }

  export function object<TPropsModel extends Record<string, Model<unknown>>>(properties: TPropsModel) {
    throw 'Not implemented';
  }
}

// 