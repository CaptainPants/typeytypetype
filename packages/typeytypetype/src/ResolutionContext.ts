import { Model } from './models/Model.js';

export interface ResolutionContext {
    getModelFromName: (name: string) => Model<unknown>;
}
