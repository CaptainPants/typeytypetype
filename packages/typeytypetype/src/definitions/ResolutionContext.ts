import { Definition } from './Definition.js';

export interface ResolutionContext {
    getDefinitionFromName: (name: string) => Definition<unknown>;
}
