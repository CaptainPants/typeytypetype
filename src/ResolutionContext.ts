import { Model } from './models/Model';

export interface ResolutionContext {
    getModelFromName: (name: string) => Model<unknown>;
}
