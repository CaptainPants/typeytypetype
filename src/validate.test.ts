import { expect, jest, test } from '@jest/globals';
import { Type } from './Type';
import { ResolutionContext } from './ResolutionContext';

test('string', () => {
    const model = Type.string();

    const resolutionContext: ResolutionContext = { getModelFromName:() => {throw new Error();} };

    expect(model.validate(resolutionContext, 'test')).toStrictEqual(true);
    expect(model.validate(resolutionContext, false)).toStrictEqual(false);
    expect(model.validate(resolutionContext, 123)).toStrictEqual(false);
    expect(model.validate(resolutionContext, {})).toStrictEqual(false);
});

test('number', () => {
    const model = Type.number();

    const resolutionContext: ResolutionContext = { getModelFromName:() => {throw new Error();} };

    expect(model.validate(resolutionContext, 123)).toStrictEqual(true);

    expect(model.validate(resolutionContext, 'test')).toStrictEqual(false);
    expect(model.validate(resolutionContext, false)).toStrictEqual(false);
    expect(model.validate(resolutionContext, {})).toStrictEqual(false);
});

test('boolean', () => {
    const model = Type.boolean();

    const resolutionContext: ResolutionContext = { getModelFromName:() => {throw new Error();} };

    expect(model.validate(resolutionContext, false)).toStrictEqual(true);
    expect(model.validate(resolutionContext, true)).toStrictEqual(true);

    expect(model.validate(resolutionContext, 'test')).toStrictEqual(false);
    expect(model.validate(resolutionContext, 123)).toStrictEqual(false);
    expect(model.validate(resolutionContext, {})).toStrictEqual(false);
});


