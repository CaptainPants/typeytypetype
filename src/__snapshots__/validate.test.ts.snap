// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object 1`] = `
"{
    "id": number;
    "name": string;
    "roles": Array<string>;
}"
`;
